[No QA] Only load full policies one-by-one as needed

roryabraham opened the pull request:
### Details
So we want to make sure that whenever a workspace or any of its sub-pages attempts to render, then we'll load the full policy for that workspace. So whenever we navigate to a workspace page (and did not just come from one), then we'll load that full policy.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/181676

### Tests
1. Sign into NewDot using an account with a workspace.
1. Go to `/settings`. Verify using logs or the network console that no full policy is loaded.
1. Click on an existing workspace. Verify using logs or the network console that the full policy is loaded.
1. Refresh the page. Verify using logs or the network console that the full policy is loaded.
1. Click on the back arrow. Verify you are taken back to the settings page.
1. Click on an existing workspace. Verify using logs or the network console that the full policy is loaded.
1. Click on a workspace sub-page. Verify using logs or the network console that no full policy is loaded.
1. Refresh the page on one of the workspace sub-pages. Verify using logs or the network console that the full policy is loaded.
1. Click on the `x` to close the modal. Verify that the full settings modal closes.


### QA Steps
Only regression tests.

### Tested On

- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

roryabraham committed:
Only load full policies when needed

roryabraham committed:
Don't create separate subnavigator for workspaces

roryabraham committed:
Fix type annotations to use strings not numbers

roryabraham committed:
Add comment linking to docs

tgolen commented on src/libs/API.js:
```diff
@@ -474,11 +474,17 @@ function GetIOUReport(parameters) {
-function GetPolicyList() {
+function GetPolicyList(policyIDList = []) {
+    if (!policyIDList || policyIDList.length > 1) {
+        LogUtil.warn('[API] Attempting to load multiple full policies at once', {policyIDList});
```
This warning doesn't really make sense to me. What is its purpose? Is it to prevent calling this on multiple policies? If so, it doesn't accomplish that at all. If that's not the purpose, then I don't really know why we want this.

roryabraham replied:
> This warning doesn't really make sense to me. What is its purpose?

The intention is indeed that the warning should be logged whenever we attempt to load multiple policies at once.

> If so, it doesn't accomplish that at all.

Sorry, I'm not seeing how it fails to accomplish that. If the `policyIDList` is falsy, then the API will attempt to load all policies for the user. If the `policyIDList.length > 1`, then the user is attempting to load multiple full policies at once. Can you help me see what I'm missing here?

iwiznia replied:
Maybe a better approach is to log a warn in prod/stg but throw on dev?

tgolen replied:
> The intention is indeed that the warning should be logged whenever we attempt to load multiple policies at once.

Haha, yes, I get that. What I mean is... who is going to look at the warning and what are they going to do about it? Just having a warning for the sake of having a warning doesn't do anybody any good.

iwiznia replied:
Umm, I am not entirely sure, but I think either warn or alrt that are logged in JS trigger a bugbot issue... (but we should check that because there's something that we ignore, it might be warns or both)

roryabraham replied:
> Maybe a better approach is to log a warn in prod/stg but throw on dev?

This makes sense, I wonder if this should be the approach for all `warn` and `alert` logs?

Also, my (unconfirmed) understanding was that an alert will trigger a bugbot issue and some >1 threshold of occurrences of a warning will trigger a bugbot issue.

roryabraham replied:
Looking [here](https://github.com/Expensify/Web-Expensify/blob/dc87516bdfa259cfd0084982d377380da8953e2e/script/bugBot.php#L356-L368) it seems the threshold is 10 occurrences in an hour. Not sure if that's warnings, alerts, or both

roryabraham replied:
And looking [here](https://github.com/Expensify/Web-Expensify/blob/b188c76fa748f334ca96f11529fa5cef89c09c48/lib/Log.php#L381-L437) it seems the behavior is the same for warnings and alerts. So >10 occurrences in an hour should create a bugbot issue.

roryabraham replied:
I updated `Log.warn` and `Log.alert` to only log server errors on staging or production. Check it out and let me know your thoughts!

iwiznia replied:
> This makes sense, I wonder if this should be the approach for all warn and alert logs?

Not sure, seems risky, but maybe would be good...  (for sure don't do that in this PR)

I was talking about this https://github.com/Expensify/Web-Expensify/blob/8d25fb6b9bcb86ab7d5a392cfaeb1ecc292bd1ca/lib/LogAPI.php#L50-L69

iwiznia replied:
> I updated Log.warn and Log.alert to only log server errors on staging or production.

Why?

roryabraham replied:
So we aren't creating bugbot issues from errors in people's local testing. I am fine removing that from this PR if you think it's risky.

roryabraham replied:
Would like to get this PR over the finish line so I just removed the `LogUtil.warn` and all the log changes for now.

iwiznia replied:
> So we aren't creating bugbot issues from errors in people's local testing. I am fine removing that from this PR if you think it's risky.

We wouldn't, we would record those errors in your local bugbot DB.

tgolen commented on src/libs/Navigation/AppNavigator/ModalStackNavigators.js:
```diff
@@ -225,7 +227,26 @@ const SettingsModalStackNavigator = createModalStackNavigator([
     },
-]);
+], {
+    state: ({data}) => {
```
Yeah, I agree that this code doesn't feel right because it's adding business logic in the wrong layer. Rory, did you try using a higher order component to share this functionality with the workspace pages?

roryabraham replied:
The HOC is a good suggestion, giving that a try!

roryabraham replied:
> As for solutions, I was thinking we would create a component that would be used by all workspace pages, enclosing the components we have now and would load the full policy then pass it down?

@iwiznia I was struggling to find a way to accomplish this with our current routing configuration, but the HOC might do it!

roryabraham replied:
Updated to use a HOC, let me know what you think! I think it's much better now, so thanks for the suggestion!

roryabraham commented on src/libs/Navigation/AppNavigator/ModalStackNavigators.js:
```diff
@@ -225,7 +227,26 @@ const SettingsModalStackNavigator = createModalStackNavigator([
+        const policyID = _.chain(data)
+            .get(['state', 'routes'], [])
+            .pluck('params')
+            .last()
```
Yeah, the StackNavigator `routes` state is like a history of routes in a stack. Navigating to a new route pushes a new route to this stack, going back pops a route.

roryabraham replied:
You can read into it more [here](https://reactnavigation.org/docs/navigation-state/)

roryabraham replied:
This code is now gone, so I'm resolving this thread.

roryabraham committed:
Merge branch 'main' into Rory-KillPolicyList

roryabraham committed:
Move policy loading logic to HOC

roryabraham committed:
Move workspace invite page to the settings modal stack

roryabraham commented:
> Something I missed in this implementation – the WorkspaceInvitePage is in it's own navigator, so if you link directly to the workspace invite page then you're not guaranteed to have a policy loaded. I think this is just an oversight and it should really be in the same navigator as the rest of the workspace pages.

I went ahead and addressed this and moved the `WorkspaceInvitePage` into the same navigator as the rest of the workspace pages.

roryabraham commented:
Updated and ready for review!

iwiznia commented on src/libs/actions/Policy.js:
```diff
@@ -41,40 +41,29 @@ function getSimplifiedEmployeeList(employeeList) {
+ * @param {String} fullPolicyOrPolicySummary.role
+ * @param {String} fullPolicyOrPolicySummary.type
+ * @param {String} fullPolicyOrPolicySummary.outputCurrency
+ * @param {Object} fullPolicyOrPolicySummary.value.employeeList
```
This is optional, right? As the summary does not have it?

roryabraham replied:
Yep, fixed!

iwiznia commented on src/pages/workspace/withFullPolicy.js:
```diff
@@ -0,0 +1,49 @@
+    };
+
+    const defaultProps = {
+        forwardedRef: () => {},
```
Can you add a comment here explaining the prop please?

roryabraham replied:
done!

roryabraham committed:
Add comment for forwardedRef and fix JSDoc

roryabraham committed:
Override Log.warn and Log.alert to only log server errors on production

roryabraham committed:
Just use local log instead of throwing an error

roryabraham committed:
Remove log changes and demote warning to hmmm

roryabraham committed:
Actually just get rid of the log entirely

roryabraham committed:
Remove extra line

roryabraham committed:
Rename API.GetPolicyList to API.GetFullPolicy

tgolen commented on src/pages/workspace/withFullPolicy.js:
```diff
@@ -0,0 +1,51 @@
+
+    const WithFullPolicy = (props) => {
+        const currentRoute = _.last(useNavigationState(state => state.routes || []));
+        const policyID = _.get(currentRoute, ['params', 'policyID'], '');
```
I checked with Marc, and we still want to standardize on using `lodashGet` everywhere (until we can _fully_ adopt underscore).

roryabraham replied:
Got rid of all the uses of `_.get`

tgolen reviewed (changes requested):
I chatted with Rory 1:1 and suggested the following changes:

1. Change `API.getPolicyList()` to only accept a single string argument for `policyID` and throw otherwise (this will remove the need to have a log)
1. Change the HOC so that it passes the policy object to the consumers (this is a more typical pattern for an HOC so it would make it more consistent)
1. Make sure the only pages that need the policy object are the ones using the HOC

roryabraham committed:
Rework HOC to pass policy as prop

roryabraham committed:
Remove console logs

roryabraham commented:
PR has been updated per @tgolen's comment above ^

tgolen commented on src/pages/workspace/WorkspaceInitialPage.js:
```diff
@@ -163,7 +151,7 @@ const WorkspaceInitialPage = ({
                                 onPress={openEditor}
                             >
-                                <Tooltip text={policy.name}>
+                                <Tooltip text={policy.name || ''}>
```
For this one, maybe it would be better to not render the tooltip if there is no name?

roryabraham replied:
Updated!

tgolen reviewed (changes requested):
Looks really good. Just this one thing I noticed.

roryabraham committed:
Don't render policy name if there is no name

iwiznia commented:
You got conflicts

Gonals commented:
Looking good overall, but there's a conflict!

roryabraham committed:
Merge branch main into Rory-KillPolicyList

roryabraham committed:
Pet-peeve comment spacing

roryabraham commented:
Conflicts resolved!

iwiznia commented on src/libs/actions/Policy.js:
```diff
@@ -144,51 +144,46 @@ function createAndNavigate(name = '') {
         .then((data) => {
             if (data.jsonCode === 200) {
-                const policy = data.policyList[0] || {};
+                const policy = lodashGet(data, 'policyList[0]', {});
```
TIL: Did not know this worked, thought you had to do `lodashGet(data, ['policyList', 0], {});`

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1463912379) to staging by @Gonals in version: 1.1.14-5 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1491852026) to production by @AndrewGable in version: 1.1.15-15 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
